### ISSUE TITLE
oracledb's Oracle_Transaction.acquireConnection ignores external connection passed in config

### ISSUE THREAD

**Issue by kenbellows** [knex/knex#3721]
Hey there! I'm writing an API using knex that queries an OracleDB instance in a way that requires me to manually interact with a `node-oracledb` connection object before kicking off a transaction. I was happy to discover that I could manually pass in a connection object that I had modified using `const trx = knex.transaction(undefined, { connection })`, which solves my problem perfectly.

However, I found that the connection object I was passing in did not seem to be the one making the db calls. For example, `node-oracledb` exposes a magic setter on `connection` objects called `connection.clientId` that sets `SYS_CONTEXT('userenv', 'client_identifier')` in the db. So I tried this test code:

```js
const knex = require('knex')({
  client: 'oracledb',
  connection: { /* ... connection info ... */ }
})

(async () => {
  const connection = await knex.client.acquireConnection()
  connection.clientId = 'testuser'
  console.log(await knex.transaction(async trx => {
    return trx.raw(`SELECT SYS_CONTEXT('userenv', 'client_identifier') FROM DUAL`)
  }, { connection })
})()
```
My expected output, what I get when I use the `node-oracledb` library directly, is:
```js
[ { "SYS_CONTEXT('userenv', 'client_identifier')": "testuser" } ]
```
My actual output was:
```js
[ { "SYS_CONTEXT('userenv', 'client_identifier')": null } ]
```
which indicates that the connection object used to make the query has no `clientId` set.

I did some digging through Knex and found the problem. There's a small bug in `lib/dialects/oracledb/transaction.js` in [the `acquireConnection()` function](https://github.com/knex/knex/blob/master/lib/dialects/oracledb/transaction.js#L53-L76):
```js
 async acquireConnection(config, cb) {
    const configConnection = config && config.connection;

    const connection = await this.client.acquireConnection();
    try {
      connection.__knexTxId = this.txid;
      connection.isTransaction = true;
      return await cb(connection);
    } finally {
      debugTx('%s: releasing connection', this.txid);
      connection.isTransaction = false;
      try {
        await connection.commitAsync();
      } catch (err) {
        this._rejecter(err);
      } finally {
        if (!configConnection) {
          await this.client.releaseConnection(connection);
        } else {
          debugTx('%s: not releasing external connection', this.txid);
        }
      }
    }
  }
```

Note the first few lines:
```js
const configConnection = config && config.connection;
const connection = await this.client.acquireConnection();
```
The `configConnection` var is set to the external connection passed in, if there is one, but that connection is then not used and a new connection is acquired.

This is the source of my problem, since my connection is indeed not being used. However, there's actually a deeper problem. The call to `this.client.acquireConnection()` pulls a connection out of knex's internally maintained connection pool, but later we get these lines:
```js
finally {
  if (!configConnection) {
    await this.client.releaseConnection(connection);
  } else {
    debugTx('%s: not releasing external connection', this.txid);
  }
}
```
This code assumes that if a `config.connection` was provided, it was used, so we should not release it back to the pool. But in fact it is not being used. So the side effect here is that when `config.connection` is passed in, not only is that connection not used, but **a connection is pulled from the pool and never released back.**

The fix is fairly straightforward; we just need to update these lines to match [the equivalent lines from the base knex Transaction class](https://github.com/kenbellows/knex/blob/master/lib/transaction.js#L210-L213):
```js
async acquireConnection(config, cb) {
    const configConnection = config && config.connection;
    const connection =
      configConnection || (await this.client.acquireConnection());
```

I actually did this locally within my `node_modules/knex` directory and my demo began working correctly, so I'm confident in the fix. I would be happy to put in a PR, but it may take me a few days before I'll have time to put together a local oracle testing environment and write tests for it.

**Comment by briandamaged** [comment]
Howdy!  Coincidentally, I'm actually in the middle of redesigning / refactoring a lot of this logic right now.  I'll actually have a PR for you to experiment with in a few minutes.  It might end up fixing your issue.

**Comment by briandamaged** [comment]
@kenbellows : Here's a PR for you to take a look at:

https://github.com/knex/knex/pull/3722

And, the associated branch:

https://github.com/briandamaged/knex/tree/feature/connection-leasing

Can you give this a try when you have a moment, plz?  Thx!

**Comment by briandamaged** [comment]
@kenbellows : I forgot to mention an important detail: this PR actually changes the `Transaction` API.  So, you will need to do the following instead:

```
const {ReuseConnection} = require('knex/lib/util/connection-leasing');


const knex = require('knex')({
  client: 'oracledb',
  connection: { /* ... connection info ... */ }
})



(async () => {
  const connection = await knex.client.acquireConnection()
  try {
    // Note: this will be passed into the call to knex.transaction(..) below
    const trxConfig = { withConnection: ReuseConnection(connection) };

    connection.clientId = 'testuser'
    console.log(await knex.transaction(async trx => {
      return trx.raw(`SELECT SYS_CONTEXT('userenv', 'client_identifier') FROM DUAL`)
    }, trxConfig)

  } finally {
    await knex.client.releaseConnection(connection);
  }



})()

```

**Comment by kenbellows** [comment]
Makes sense to me! I'm having some environment issues and I don't think I'll be able to test it myself in the short term, but I've read the changes and it looks like it should solve the issue nicely!

Any idea when the next release will be?

**Comment by kenbellows** [comment]
I see you all are preparing for the 1.0.0 release! Congrats, that's a big milestone! Is there any ETA yet for when that release will likely happen? Since it's a release that introduces breaking changes, such as the API changes you're suggesting here, would it make sense to also make a change in the 0.x code to fix this bug for those who cannot immediately upgrade?

**Comment by briandamaged** [comment]
Howdy again!  I'm not sure about the official release timeline (I'm rather new to the team).  As for the refactoring that I've been working on: at the moment, I think we're still trying to decide which approach we want to take.  Ex: the current code fixes certain things, but also breaks backwards-compatibility in various ways.  So, we're still drafting a few proposals.

As for fixing that specific bug: that might be worthwhile.  Do you already have a PR for the change?  If not, then I can probably create one in a little bit.

**Comment by kenbellows** [comment]
The code change is trivial, it's like 3 lines, but I'm not confident about writing the tests and such; I started setting up a local environment last night but ran into some issues on my machine, and I don't know that I have time to dig in and figure out what's wrong. So if I can submit a PR with just the code changes, I'll have that done today, but if I need to write tests and everything to get the PR submitted, then I'd happily accept the help :)

**Comment by briandamaged** [comment]
@kenbellows : Sry -- I had to take care of a few other things, so it took me a bit to get back to this.  In either case, I've created a PR here:

#3731

**Comment by kibertoad** [comment]
I'm thinking about cutting new 0.20 release tomorrow. @briandamaged @maximelkin @lorefnon Anything you would like merged prior to that?

**Comment by briandamaged** [comment]
@kibertoad : Nothing needs merging on my side.  I'm still just hammering out details for some experimental refactorings.  Thx for checking!

**Comment by kenbellows** [comment]
With #3731 merged in, this issue is resolved! Thanks @briandamaged for jumping on it so quickly! Any more movement on that 0.20 release?

**Comment by kibertoad** [comment]
@kenbellows I'll release a version shortly, sorry for delay.

**Comment by kibertoad** [comment]
@kenbellows ReleaseD in 0.20.12
